### ISSUE TITLE
userfaultfd async wp mode in page guard manager

### ISSUE THREAD

**Issue by ishitatsuyuki** [LunarG/gfxreconstruct#1460]
Congratulations on shipping userfaultfd support in the page guard manager!

It looks like overhead was noted as a concern when using the userfaultfd option. I would like to point out that a recently added "Async Write-Protect" mode for userfaultfd is intended to address change-tracking use cases like this.

The feature is documented at https://docs.kernel.org/admin-guide/mm/userfaultfd.html#write-protect-notifications, and is used in Valve's Wine fork (https://github.com/ValveSoftware/wine/blob/bleeding-edge/dlls/ntdll/unix/virtual.c) to emulate Win32's write watches. What gfxr is doing should be similar to write watches, so perhaps the Wine source can be used as a reference when implementing async wp support.

**Comment by panos-lunarg** [comment]
Hi @ishitatsuyuki

Thank you for taking the time to suggest possible improvements.
From what I understand you are referring to `UFFDIO_WRITEPROTECT_MODE_WP` feature flag. We are aware of that but there are some limitations with it and it turns out that it's not so useful for our case.

The first reason is that this feature is relatively new and is not broadly supported by the majority of kernels that exist now on Android phones.

The second and more important reason is that it doesn't do exactly what we need.
From what I understand `UFFDIO_WRITEPROTECT_MODE_WP` should enable to track writes even to allocated/existing pages. Unfortunately this is not enough in our case as we also want to track reads, not only writes.

I will take a look at the source file you pointed. Perhaps there is something clever there that we can also implement.

**Comment by ishitatsuyuki** [comment]
You are right that async WP doesn't provide any support for detecting read-only page faults.

My understanding is that async WP should eliminate the need for shadow memory and hence it will be unnecessary to track read-only faults, like how the D3D12 backend works when using GetWriteWatch. I haven't looked at the implementation yet, but please let me know in case the Vulkan backend works in some different way such that skipping tracking reads would not be viable.

**Comment by panos-lunarg** [comment]
To be honest we haven't tried skipping shadow memory completely with userfaultfd (or maybe we did but that was some time ago and I can't remember exactly).
We have tried that with the `mprotect` mechanism (apply the `mprotect` + `SIGSEGV` trick directly on the mapped memory returned by the driver) but it turned out that it didn't work out well. IIRC there were random crashes.
I guess it wouldn't hurt to try it with userfaultfd.

**Comment by panos-lunarg** [comment]
I am afraid that this is not possible. If I understand the uffd documentation correctly, both `UFFDIO_REGISTER_MODE_MISSING` and `UFFDIO_REGISTER_MODE_WP` must be applied on private anonymous memory regions. The memory regions returned from the driver are not expected to fulfill this criteria.
I did some very brief tests by not allocating a shadow memory and registering the mapped memory directly to uffd and it seems that the registration fails:
```
E gfxrecon: ioctl/uffdio_register: Invalid argument
E gfxrecon: uffdio_register.range.start: 0x7700769000
E gfxrecon: uffdio_register.range.len: 524288
```
Checking the /proc/pid/maps file for this region:
```
7700769000-77007e9000 rw-s 10e10e000 00:10 950                           /dev/dri/renderD128
```

**Comment by ishitatsuyuki** [comment]
This is unfortunately poorly documented but any memory region is allowed if you use ASYNC WP. Could this help?

https://github.com/torvalds/linux/blob/6d69b6c12fce479fde7bc06f686212451688a102/include/linux/userfaultfd_k.h#L225

**Comment by panos-lunarg** [comment]
Interesting. This doesn't agree with the documentation found online. I only tried applying `UFFDIO_REGISTER_MODE_MISSING` that is currently used. I'll look into it a bit more

**Comment by panos-lunarg** [comment]
No it's still failing. I'm initializing with `UFFD_FEATURE_PAGEFAULT_FLAG_WP`, registering only with `UFFDIO_REGISTER_MODE_WP` and registrations fail with `EINVAL`. Tried this on both android and desktop. Maybe I am missing something? Have you ever done something similar?

**Comment by panos-lunarg** [comment]
I recently installed Ubuntu 24.04 which has 6.8 kernel with support for the `UFFD_FEATURE_WP_ASYNC` feature and did a quick prototype to see if it would work with GFXR.

Unfortunately it seems that problem that the memory returned by the driver cannot be write protected persists even with the new feature. This is probably because the drivers (in general) already apply some form of protection the the mapped memory so applying another protection causes problems.

I tested this on a desktop machine with the radv driver. Protecting the memory with `UFFDIO_WRITEPROTECT` causes the application to hang on the first attempt to write. Applying the protection with `PAGEMAP_SCAN` ioctl fixes the hang but in the end no writes are detected. (This was a bit unexpected since [the documentation I read](https://docs.kernel.org/admin-guide/mm/pagemap.html#pagemap-scan-ioctl) states that these actions do the same thing)

Prototype branch: https://github.com/LunarG/gfxreconstruct/commit/6c135837e1bbe85075591dcd1dc7269a9af0c5bf

**Comment by per-mathisen-arm** [comment]
I've also noticed similar problems with multiple drivers (intel, nvidia and mali, from memory). From talking to driver people, this is apparently not trivial to fix on linux due to a kernel imitation/bug. It might be easier to try to piggy-back on the driver protection and its write tracking through a Vulkan extension...

**Comment by panos-lunarg** [comment]
> I've also noticed similar problems with multiple drivers (intel, nvidia and mali, from memory). From talking to driver people, this is apparently not trivial to fix on linux due to a kernel imitation/bug. It might be easier to try to piggy-back on the driver protection and its write tracking through a Vulkan extension...

I am not familiar with the kernel's internals and what is the limitation in this case, but as far as I can tell, an extension currently seems the only solution for Linux
